### PR TITLE
updated the last unit test; added features for patch and cell labelin…

### PR DESCRIPTION
**Pull request by Jurwen**
## Fix for the previous unit test / Function labels for the material interface

* Updated the function file and unit test files for local value change.
* Changed the patch-to-function label into two flags. For material interface, the first flag is the function index on the positive side, and the second flag is on the negative side. For IA, only the first flag is activated. Adjusted the related API. 
* Added cell-to-function label for material interface. It's using one `size_t` to represent the function index per cell.
* Updated IO to separated output IA and MI results as they become different.

### CODE REVIEW

**Comment by Jurwen** [comment]
* Additional unit tests for MI. Sphere examples become voronoi cells in MI. As MI only needs to output one flag per cell regardless of any function that does not dominate any area in the space, the two last two unit tests become trivial in MI, thus removed. The configuration of all unit tests is the following graph: 

Notice that the above cell indices have only 1 digit presenting the function index that dominates this cell.

image

**Comment by qnzhou** [comment]
Thanks Owen! I do not understand this visualization:
image
I expect a perfectly flat plane. The colors of the plane looks weird in this example.

Can you add the following unit tests:
1. Compute the material of a single function. It should produce a single cell with the function label of this function.
2. Compute the material interface of 3 nested spheres of different radius. I.e. one function is larger than the other two at all points.
3. Compute the material interface of 8 spheres of the same radius centered at the 8 corners of a cube. Check the number of non-manifold vertices (should be 1).

**Comment by Jurwen** [comment]
Here is the same visualization from a different angle. 

image

The plane isn't perfect. It has bumps so that the lighting makes the previous visualization weird. 

For the additional unit test: 
* There is no cell output from the first and third cases. The cell is computed from the patches, and when there's no patch, there is no cell. This is an edge case that the algorithm does not cover. 
* Will add this edge case in so that when there is only one cell with no other meta data, we will still output this cell with a function label.

**Comment by Jurwen** [comment]
### Unit Test with 8 Spheres:

* There is significant topological error using the uniform grid as the simplicial elements are too large at function intersections. 

<div class="container">
  image
image
</div>

* Thus switching to grids generated by our adaptive grid generation code. Topologically, the result gets better and the meta data count looks correct.

<div class="container">
image

image
</div>

* There are minor issues with an edge case (I found along the way) which triggers an exception from Charles' code: "More than one tets containing the iso-faces incident to the iso-edge." where in above example with degenerate sphere locations. If their intersection coincides with boundary of the tetrahedrons, then such exception triggers. My guess is when an iso-face is shared by different tetrahedrons when iso-faces and tetrahedron faces coincide. I don't know if we should be concerned about this case. Currently, the unit test is constructed such that the spheres are not in the degenerate position (exact corners of the gird).

**Comment by qnzhou** [comment]
> There is significant topological error using the uniform grid as the simplicial elements are too large at function intersections. ... Thus switching to grids generated by our adaptive grid generation code. Topologically, the result gets better and the meta data count looks correct.

Interesting! Good to know. Do you need to additional tuning to get this case to work with our adaptive grid?

> There are minor issues with an edge case (I found along the way) which triggers an exception from Charles' code: "More than one tets containing the iso-faces incident to the iso-edge." where in above example with degenerate sphere locations.

Can you also add this as a separate unit test case? You can use the `[.may_fail]` tag to indicate this test may fail.

**Comment by Jurwen** [comment]
> Do you need to additional tuning to get this case to work with our adaptive grid?

Sort of, as these spheres have identical opposite gradients at intersection locations. I have to limit the tetrahedron number to  terminate the subdivision. 

> Can you also add this as a separate unit test case? You can use the `[.may_fail]` tag to indicate this test may fail.

Will do.

**Comment by qnzhou** [comment]
> Sort of, as these spheres have identical opposite gradients at intersection locations. I have to limit the tetrahedron number to terminate the subdivision.

Let's remember this case and come back to it when we work on the other codebase.

File: tests/test_implicit_networks.cpp [hunk 1]
```diff
@@ -680,8 +680,88 @@ TEST_CASE("material interface on known examples", "[MI][examples]") {
...
-    load_tet_mesh(std::string(TEST_FILE) + "/mesh.json", pts, tets);
     SECTION("eight sphere") {
+        load_tet_mesh(std::string(TEST_FILE) + "/mesh.json", pts, tets);
+        // compute function values on tet grid vertices
+        size_t n_pts = pts.size();
+        size_t n_func = 8;
+        funcVals.resize(n_pts, n_func);
+        size_t func_id;
+        
+        if (!load_functions(std::string(TEST_FILE) + "/8-sphere.json", pts, funcVals)) {
```
Comment (qnzhou) [inline tests/test_implicit_networks.cpp:691]: Please checkin this file.

File: src/material_interface.h [hunk 2]
```diff
@@ -6,7 +6,35 @@
 #define ROBUST_IMPLICIT_NETWORKS_MATERIAL_INTERFACE_H
 
 #include "robust_implicit_networks.h"
-
+///
+/// The body of computing the implicit arrangement; used in `src/material_interface.cpp`
```
Comment (qnzhou) [inline src/material_interface.h:10]: "implicit arrangement" -> "material interface"

File: src/material_interface.h [hunk 3]
```diff
@@ -6,7 +6,35 @@
...
+///
+///  @param[out] MI_pts            Vertices at the surface network mesh
+///  @param[out] MI_faces          Polygonal faces at the surface network mesh
+///  @param[out] patches            A connected component of faces bounded by non-manifold edges
+///  @param[out] patch_function_label           a pair of two indices of function that are domiating the neighboring cells of this patch
+///  @param[out] MI_edges          Edges at the surface network mesh
+///  @param[out] chains         Chains of non-manifold edges
+///  @param[out] non_manifold_edges_of_vert         Indices of non-manifold vertices
+///  @param[out] shells         A connected component of the boundary partitioned by the surface network; encoded by patch indices: positive patch i -> 2i, negative patch i->2i+1
+///  @param[out] arrangement_cells          A 3D region partitioned by the surface network; encoded by a vector of shell indices
```
Comment (qnzhou) [inline src/material_interface.h:29]: Please update this line for material interface.

File: src/material_interface.h [hunk 4]
```diff
@@ -6,7 +6,35 @@
...
+///
+///
+///  @param[out] MI_pts            Vertices at the surface network mesh
+///  @param[out] MI_faces          Polygonal faces at the surface network mesh
+///  @param[out] patches            A connected component of faces bounded by non-manifold edges
+///  @param[out] patch_function_label           a pair of two indices of function that are domiating the neighboring cells of this patch
+///  @param[out] MI_edges          Edges at the surface network mesh
+///  @param[out] chains         Chains of non-manifold edges
+///  @param[out] non_manifold_edges_of_vert         Indices of non-manifold vertices
+///  @param[out] shells         A connected component of the boundary partitioned by the surface network; encoded by patch indices: positive patch i -> 2i, negative patch i->2i+1
```
Comment (qnzhou) [inline src/material_interface.h:28]: I would rephrase this as 

> An array of shells. Each shell is a connected component of patches. Even patch index, `2*i`, indicates patch `i` is consistently oriented with the shell. Odd patch index, `2*i+1`, indicates patch `i` has opposite orientation with respect to the shell.

File: src/material_interface.h [hunk 5]
```diff
@@ -6,7 +6,35 @@
...
+///  @param[out] arrangement_cells          A 3D region partitioned by the surface network; encoded by a vector of shell indices
+///  @param[out] cell_function_label            a 1D vector of `size_t` of values of function index for each cell.
```
Comment (qnzhou) [inline src/material_interface.h:30]: Please update this doc for material interface.

File: src/material_interface.cpp [hunk 6]
```diff
@@ -614,6 +615,12 @@ bool material_interface(
                 material_cells.emplace_back(1);
                 material_cells.back()[0] = i;
             }
+            // resolve degenerate cases where all materials are dominated, i.e., one cell bounded by boundaries.
```
Comment (qnzhou) [inline src/material_interface.cpp:618]: I think the degenerate case is when none of the material interface intersect with the bounding box of our grid. Only 1 material dominates within the grid bounding box.

File: src/material_interface.cpp [hunk 7]
```diff
@@ -614,6 +615,12 @@ bool material_interface(
                 material_cells.emplace_back(1);
                 material_cells.back()[0] = i;
```
Comment (qnzhou) [inline src/material_interface.cpp:616]: Do we have information about the orientation of a shell with respect to a material cell? Can a shell be shared by two neighboring material cells?

File: src/mesh_connectivity.cpp [hunk 8]
```diff
@@ -476,3 +479,40 @@ std::vector<std::vector<bool>> sign_propagation(const std::vector<std::vector<si
...
+std::vector<size_t> sign_propagation_MI(const std::vector<std::vector<size_t>>& material_cells,
+                      const std::vector<size_t>& shell_of_half_patch,
+                      const std::vector<std::vector<size_t>>& shells,
+                      const std::vector<std::pair<size_t, size_t>>& patch_function_label,
+                      size_t n_func,
+                      const std::vector<double>& sample_function_label){
+    std::vector<size_t> cell_function_label(material_cells.size(), Mesh_None);
+    for (size_t cell_index = 0; cell_index < material_cells.size(); cell_index++){
+        for (auto shell : material_cells[cell_index]){
+            /// resolve degenerate cases where all materials are dominated, i.e., the cell contains `Mesh_None` as its element.
```
Comment (qnzhou) [inline src/mesh_connectivity.cpp:492]: "all materials are dominated" is not an accurate description of this degenerate case.

**Review by qnzhou — COMMENTED** [review]
Thanks Owen! I have a few more comments.

**Review by qnzhou — APPROVED** [review]
